hamilton/function_modifiers/recursive.py

```diff
@@ -626,3 +629,314 @@ def prune_nodes(nodes: List[node.Node], select: Optional[List[str]] = None) -> L
+    ) -> node.Node:
+        "Node that adds to / overrides columns for the original dataframe based on selected output."
+        if self.is_async:
+
+            async def new_callable(**kwargs) -> Any:
+                df = kwargs[upstream_node]
+                columns_to_append = {}
+                for column in self.select:
+                    columns_to_append[column] = kwargs[column]
+                new_df = registry.with_columns(df, columns_to_append)
+                return new_df
+        else:
+
+            def new_callable(**kwargs) -> Any:
+                df = kwargs[upstream_node]
+                columns_to_append = {}
+                for column in self.select:
+                    columns_to_append[column] = kwargs[column]
+
+                new_df = registry.with_columns(df, columns_to_append)
+                return new_df
+
+        column_type = registry.get_column_type_from_df_type(dataframe_type)
+        input_map = {column: column_type for column in self.select}
+        input_map[upstream_node] = dataframe_type
+
+        return node.Node(
+            name=node_name,
+            typ=dataframe_type,
+            callabl=new_callable,
+            input_types=input_map,
+        )
+
+    def inject_nodes(
+        self, params: Dict[str, Type[Type]], config: Dict[str, Any], fn: Callable
+    ) -> Tuple[List[node.Node], Dict[str, str]]:
+        self.is_async = inspect.iscoroutinefunction(fn)
+        namespace = fn.__name__ if self.namespace is None else self.namespace
+
+        inject_parameter, initial_nodes, dataframe_type = self._get_inital_nodes(
+            fn=fn, params=params
+        )
+
+        subdag_nodes = subdag.collect_nodes(config, self.subdag_functions)
+
+        # TODO: for now we restrict that if user wants to change columns that already exist, he needs to
+        # pass the dataframe and extract them himself. If we add namespace to initial nodes and rewire the
+        # initial node names with the ongoing ones that have a column argument, we can also allow in place
+        # changes when using columns_to_pass
+        if with_columns._check_for_duplicates(initial_nodes + subdag_nodes):
+            raise ValueError(
+                "You can only specify columns once. You used `columns_to_pass` and we "
+                "extract the columns for you. In this case they cannot be overwritten -- only new columns get "
+                "appended. If you want to modify in-place columns pass in a dataframe and "
+                "extract + modify the columns and afterwards select them."
+            )
+
+        pruned_nodes = prune_nodes(subdag_nodes, self.select)
+        if len(pruned_nodes) == 0:
+            raise ValueError(
+                f"No nodes found upstream from select columns: {self.select} for function: "
+                f"{fn.__qualname__}"
+            )
+        # In case no node is selected we append all possible nodes that have a column type matching
```

Comment:
Adding here another comment for future reference:

At the moment `ResultBuilder` will create a column out of a scalar, native Polars `with_columns` does the same.

In our `with_columns` implementation this results in an error since we enforce only matching series type can be appended into the dataframe.

I am torn how we want to handle this: on the one hand we should try to mimick the same behaviour  as external libraries (since that is what users are used to / easier adoption with legacy code), on the other hand enabling this can lead to nasty bugs because it will fail silently.

Reply:
To circle back, I thnk we're removing result builder to reduce complexity, right?